scheduler/src/cook/quota.clj

```diff
@@ -142,22 +143,70 @@
+        query '[:find ?u ?a
+                :in $ ?t ?pool-name ?requesting-default-pool
+                :where
+                [?e :quota/user ?u]
+                [?e :quota/resource ?r]
+                [?r :resource/type ?t]
+                [?r :resource/amount ?a]
+                [(cook.pool/check-pool $ ?r :resource/pool ?pool-name ?requesting-default-pool)]]]
+    (->> (d/q query db type pool-name' requesting-default-pool)
+      (into {}))))
+
+(defn- retrieve-count-quota-from-user-entities
+  "Returns the user->count-quota for all count quota specified on the user entity."
+  [db users]
+  (let [query '[:find ?u ?c
+                :in $ [?u ...]
+                :where
+                [?e :quota/user ?u]
+                [?e :quota/count ?c]]]
+    (->> (d/q query db users)
+      (into {}))))
+
+(defn- retrieve-user->count-quota
+  "Returns the map for user to count-quota for all the provided users."
+  [db pool-name all-users default-quota]
+  (let [user->resource-count-quota (pool-name->type->user->quota db pool-name :count)
+        remaining-users (vec (set/difference (set all-users) (set (keys user->resource-count-quota))))
+        user->entity-count-quota (when (seq remaining-users)
+                                   (retrieve-count-quota-from-user-entities db remaining-users))]
+    (pc/map-from-keys
+      (fn [user]
+        (int
+          ; prefer resource over the field on the user for count quota
+          (or (get user->resource-count-quota user)
+              (get user->entity-count-quota user)
+              default-quota)))
+      all-users)))
+
 (defn create-user->quota-fn
   "Returns a function which will return the quota same as `(get-quota db user)`
    snapshotted to the db passed in. However, it queries for all users with quota
    and returns the `default-user` value if a user is not returned.
    This is usefully if the application will go over ALL users during processing"
   [db pool-name]
-  (let [all-quota-users (d/q '[:find [?user ...]
-                               :where
-                               [?q :quota/user ?user]]
-                             db)
-        user->quota-cache (->> all-quota-users
-                               (map (fn [user]
-                                      [user (get-quota db user pool-name)]))
-                               ;; In case default-user doesn't have an explicit quota
-                               (cons [default-user (get-quota db default-user pool-name)])
-                               (into {}))]
+  (let [default-type->quota (get-quota db default-user pool-name)
+        default-count-quota (get default-type->quota :count)
+        all-resource-types (util/get-all-resource-types db)
+        type->user->quota (pc/map-from-keys #(pool-name->type->user->quota db pool-name %) all-resource-types)
+        all-quota-users (d/q '[:find [?user ...] :where [?q :quota/user ?user]] db)
```

Comment:
Probably worth a clojuredoc comment that there are no duplicates in this list.

Reply:
Added comment.